GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,258 @@
+        /// <param name="path">The path being trimmed</param>
+        /// <param name="directoryTarget">The directory target whose path to trim from the path</param>
+        /// <returns>The newly formatted path with the directory trimmed</returns>
+        private string TrimDirectoryFromPath(string path, string directoryTarget)
+        {
+            return path.Substring(directoryTarget.Length);
+        }
+
+        private void IncreaseDictionaryCounterByKey(Dictionary<string, int> countingDictionary, string key)
+        {
+            if (!countingDictionary.TryGetValue(key, out int count))
+            {
+                count = 0;
+            }
+
+            countingDictionary[key] = ++count;
+        }
+
+        private decimal CalculateHealthMetric(int hydratedFileCount, int totalFileCount)
+        {
+            if (totalFileCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)hydratedFileCount / (decimal)totalFileCount;
+        }
+
+        public class GVFSEnlistmentPathData
+        {
+            public List<string> GitFolderPaths;
+            public List<string> GitFilePaths;
+            public List<string> PlaceholderFolderPaths;
+            public List<string> PlaceholderFilePaths;
+            public List<string> ModifiedFolderPaths;
+            public List<string> ModifiedFilePaths;
+
+            public GVFSEnlistmentPathData()
+            {
+                this.GitFolderPaths = new List<string>();
+                this.GitFilePaths = new List<string>();
+                this.PlaceholderFolderPaths = new List<string>();
+                this.PlaceholderFilePaths = new List<string>();
+                this.ModifiedFolderPaths = new List<string>();
+                this.ModifiedFilePaths = new List<string>();
+            }
+
+            public void AddSkipWorkTreeFilePaths(List<string> skipWorkTreeFilePaths)
+            {
+                this.ModifiedFilePaths.Union(skipWorkTreeFilePaths);
+            }
+
+            public void NormalizeAllPaths()
+            {
+                this.NormalizePaths(this.GitFolderPaths);
+                this.NormalizePaths(this.GitFilePaths);
+                this.NormalizePaths(this.PlaceholderFolderPaths);
+                this.NormalizePaths(this.PlaceholderFilePaths);
+                this.NormalizePaths(this.ModifiedFolderPaths);
+                this.NormalizePaths(this.ModifiedFilePaths);
+            }
+
+            private void NormalizePaths(List<string> paths)
+            {
+                for (int i = 0; i < paths.Count; i++)
+                {
+                    paths[i] = paths[i].Replace('\\', GVFSConstants.GitPathSeparator);
+                    paths[i] = paths[i].TrimStart(GVFSConstants.GitPathSeparator);
+                    paths[i] = paths[i].TrimEnd(GVFSConstants.GitPathSeparator);

```

Comment:
Potential performance improvement (no need to block the PR on this comment):

Reduce the number of string allocations by using `Substring` here instead of `TrimStart` + `TrimEnd`.

Reply:
Is there a substring method which only takes the substring based on a condition? I use `Trim..()` here because if the character is there it gets trimmed, and if it isn't then nothing happens.

Reply:
> Is there a substring method which only takes the substring based on a condition? I

Not that I am aware of.  To move to `Substring` you'd need to check the start and end of the string yourself to calculate the appropriate starting index and length.

I think the code you have now is much more readable than that approach (only worth investigating the `Substring` if you find there are perf issues here)

Reply:
I missed @kewillford's earlier comment and agree that we can simply use `Trim`